### ISSUE TITLE
Realizing a tensor influences another already realized tensor.

### ISSUE THREAD

**Issue by StanislavNikolov** [geohot/tinygrad#10640]
Take this example
```python
from tinygrad import Tensor, nn

Tensor.manual_seed(13)
Tensor.training = True

class Autoencoder():
    def __init__(self):
        self.e1 = nn.Conv2d(3, 32, 3, stride=2)
        self.d1 = nn.ConvTranspose2d(32, 3, 3, stride=2)
    def __call__(self, x: Tensor) -> Tensor:
        return self.d1(self.e1(x).leaky_relu()).interpolate((1, 3, 512,512))

ae = Autoencoder()
opt = nn.optim.Adam(nn.state.get_parameters(ae))
x = Tensor.rand((1, 3, 512, 512))

opt.zero_grad()
(x-ae(x)).square().mean().backward()

oldv = opt.v[0].flatten()[0].item() # Initially 0

# optim.LAMB does self.v[i].assign(self.b2 * self.v[i] + (1.0 - self.b2) * (g * g))
g = opt.params[0].grad
newv = (opt.b2 * opt.v[0]) + (1.0 - opt.b2) * (g * g)
newv = newv.numpy()

_ = opt.schedule_step()

opt.v[0].realize()

# Old=0, opt.v[0] = newv - OK
print(oldv, opt.v[0].flatten()[0].item(), newv.flatten()[0])

# Reaizing m[0] should have no effect on v[0]
opt.m[0].realize()

# opt.v[0] changed?!
print(oldv, opt.v[0].flatten()[0].item(), newv.flatten()[0])
```

`opt.m[0].realize()` changed the value of `opt.v[0]`, although it was already realized earlier. My understanding is that this should not happen.

**Comment by StanislavNikolov** [comment]
Before you say that calling `schedule_step()` manually and not realizing all of them with `Tensor.realize(*_)` is weird - I was debugging this issue in a large codebase and this is the smallest code sample I could reasonably cut out.

**Comment by chenyuxyz** [comment]
the issue seems to be `schedule_step` is not just scheduling and it has side effects. As you have pointed out, realizing the schedule output would not have this issue

i don't have a good solution to update `schedule_step`. Do you also manually calling `schedule_step` in the large codebase?

**Comment by StanislavNikolov** [comment]
I was wrong, even just using opt.step() also has this issue:
```py
from tinygrad import Tensor, nn

Tensor.manual_seed(13)
Tensor.training = True

class Autoencoder():
    def __init__(self):
        self.e1 = nn.Conv2d(3, 32, 3, stride=2)
        self.d1 = nn.ConvTranspose2d(32, 3, 3, stride=2)
    def __call__(self, x: Tensor) -> Tensor:
        return self.d1(self.e1(x).leaky_relu()).interpolate((1, 3, 512,512))

ae = Autoencoder()
opt = nn.optim.Adam(nn.state.get_parameters(ae))
x = Tensor.rand((1, 3, 512, 512))

opt.zero_grad()
(x-ae(x)).square().mean().backward()

oldv = opt.v[0].flatten()[0].item() # Initially 0, OK

# optim.LAMB does self.v[i].assign(self.b2 * self.v[i] + (1.0 - self.b2) * (g * g))
g = opt.params[0].grad
newv = (opt.b2 * opt.v[0]) + (1.0 - opt.b2) * (g * g)
newv = newv.numpy()

opt.step()

# opt.v[0] is not (1-0.999) * g * g?!
print(oldv, opt.v[0].flatten()[0].item(), newv.flatten()[0])
```
You say a potential solution is not use the optimizer at all, but I believe there is a bug in the core of tinygrad, since the optimizer's code is short and simple and I don't see anything obviously wrong with it after poking for a few hours.

**Comment by wozeparrot** [comment]
if you add a `.realize()` to `_new_optim_param` it fixes:

```py
  def _new_optim_param(self) -> list[Tensor]:
    if self.fused: return [Tensor.zeros(self.pos_params[-1], dtype=dtypes.float32, device=self.device, requires_grad=False).contiguous()]
    return [Tensor.zeros(*t.shape, dtype=dtypes.float32, device=t.device, requires_grad=False).contiguous().realize() for t in self.params]
```

probably a scheduler thing @Qazalin

EDIT: might be the same as or similar to #4562

**Comment by Qazalin** [comment]
It is the same problem @wozeparrot. https://github.com/tinygrad/tinygrad/pull/10647 fixes this issue.

It's a fundamental problem with CONST, even when contiguous is applied, CONTIGUOUS(CONST) is a singleton across the entire program. When it is realized in-place, it becomes unique Buffer, which is safe to later assign to.

Multiple parts of the code can assign to the same CONTIGUOUS(CONST) Tensor, and you'd end up with different values.

**Comment by StanislavNikolov** [comment]
I confirm https://github.com/tinygrad/tinygrad/pull/10647 fixes the issue. Amazing response time :D

**Comment by chenyuxyz** [comment]
fixed now
